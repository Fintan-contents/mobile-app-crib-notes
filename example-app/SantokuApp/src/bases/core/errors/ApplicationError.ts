/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {ErrorWrapper} from './ErrorWrapper';

export class ApplicationError extends ErrorWrapper {
  constructor();
  constructor(cause: unknown);
  constructor(message: string);
  constructor(cause: unknown, errorCode?: string);
  constructor(message: string, errorCode?: string);
  constructor(message: string, cause: unknown, errorCode?: string);
  constructor(messageOrCause?: unknown, causeOrErrorCode?: unknown, errorCode?: string) {
    if (typeof messageOrCause === 'string') {
      if (typeof causeOrErrorCode === 'string' && !errorCode) {
        super(messageOrCause);
        this._errorCode = causeOrErrorCode;
      } else {
        super(messageOrCause, causeOrErrorCode);
        this._errorCode = errorCode;
      }
    } else {
      super(messageOrCause);
      if (typeof causeOrErrorCode === 'string') {
        this._errorCode = causeOrErrorCode;
      }
    }
  }

  private readonly _errorCode?: string;

  get errorCode() {
    return this._errorCode;
  }
}

export function isApplicationError(error?: unknown): error is ApplicationError {
  return error != null && typeof error === 'object' && error instanceof ApplicationError;
}
