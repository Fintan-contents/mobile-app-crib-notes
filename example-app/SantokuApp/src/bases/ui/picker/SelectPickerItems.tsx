import {PickerProps} from '@react-native-picker/picker';
import {log} from 'bases/logging';
import React from 'react';
import {Platform, StyleProp, TextStyle} from 'react-native';

import {Item} from './SelectPicker';

export type SelectPickerItemsProps<ItemT> = Omit<
  PickerProps<React.Key | ItemT>,
  'enabled' | 'mode' | 'prompt' | 'dropdownIconColor' | 'dropdownIconRippleColor' | 'onFocus' | 'onBlur' | 'style'
> & {
  items: Item<ItemT>[];
  keyExtractor?: ((item: Item<ItemT>, index: number) => string) | undefined;
  /**
   * Pickerのスタイル
   * @platform ios
   */
  style?: StyleProp<TextStyle>;
  /**
   * アイテムの高さ
   * @platform android
   */
  itemHeight?: number;
  /**
   * 選択状態のアイテムのテキストカラー
   * @platform android
   */
  activeColor?: string;
  /**
   * 未選択状態のアイテムのテキストカラー
   * @platform android
   */
  inactiveColor?: string;
  /**
   * FlatListのtestID
   * @platform android
   */
  flatListTestID?: string;
  /**
   * SeparatorのtestID
   * @platform android
   */
  separatorTestID?: string;
  /**
   * SelectPickerItemのPressableのtestIDプレフィックス
   * @platform android
   */
  itemPressableTestIDPrefix?: string;
  /**
   * SelectPickerItemのTextのtestIDプレフィックス
   * @platform android
   */
  itemTextTestIDPrefix?: string;
};

export const SelectPickerItems = <ItemT extends unknown>(props: SelectPickerItemsProps<ItemT>) => {
  React.useEffect(() => {
    log.warn(`SelectPickerItems is not supported on: ${Platform.OS}`);
  }, []);
  return null;
};
