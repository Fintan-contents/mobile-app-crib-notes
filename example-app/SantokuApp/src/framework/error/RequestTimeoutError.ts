export class RequestTimeoutError extends Error {}
