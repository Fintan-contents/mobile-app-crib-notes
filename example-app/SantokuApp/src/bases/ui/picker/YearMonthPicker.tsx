import React from 'react';
import {
  Pressable,
  StyleSheet,
  Text,
  TextInput,
  TextInputProps as RNETextInputProps,
  View,
  ViewProps,
} from 'react-native';

import {DefaultPickerAccessory, DefaultPickerAccessoryProps} from './DefaultPickerAccessory';
import {PickerBackdrop, PickerBackdropProps} from './PickerBackdrop';
import {PickerContainer, PickerContainerProps} from './PickerContainer';
import {SelectPickerItems, SelectPickerItemsProps} from './SelectPickerItems';
import {RequiredYearMonth, YearMonth} from './YearMonth';
import {useYearMonthPicker} from './useYearMonthPicker';

type TextInputProps = Omit<RNETextInputProps, 'value' | 'editable'>;
export type YearMonthPickerProps = {
  /**
   * 選択された年月
   */
  selectedValue: YearMonth | undefined;
  /**
   * 選択可能な最大日付
   */
  maximumYearMonth: RequiredYearMonth;
  /**
   * 選択可能な最小日付
   */
  minimumYearMonth: RequiredYearMonth;
  /**
   * 年のサフィックス
   */
  yearSuffixLabel?: string;
  /**
   * 月のサフィックス
   */
  monthSuffixLabel?: string;
  /**
   * アイテムが選択された場合に呼び出される関数
   */
  onSelectedItemChange?: (yearMonth?: YearMonth) => void;
  /**
   * PickerBackdropをタップして閉じた場合に呼び出される関数
   */
  onDismiss?: (yearMonth?: YearMonth) => void;
  /**
   * DeleteLabelがタップされた場合に呼び出される関数
   * タップ後、YearMonthPickerは自動で閉じます。
   */
  onDelete?: (yearMonth?: YearMonth) => void;
  /**
   * CancelLabelがタップされた場合に呼び出される関数
   * タップ後、YearMonthPickerは自動で閉じます。
   */
  onCancel?: (yearMonth?: YearMonth) => void;
  /**
   * DoneLabelがタップされた場合に呼び出される関数
   * タップ後、YearMonthPickerは自動で閉じます。
   */
  onDone?: (yearMonth?: YearMonth) => void;
  /**
   * プレースホルダ
   */
  placeholder?: string;
  /**
   * 選択されたアイテムを表示するテキストコンポーネントのProps
   */
  textInputProps?: TextInputProps;
  /**
   * Pickerコンポーネントを囲むContainerのProps
   */
  pickerItemsContainerProps?: ViewProps;
  /**
   * PickerのProps
   */
  pickerProps?: Omit<SelectPickerItemsProps<YearMonth>, 'items' | 'selectedValue' | 'onValueChange' | 'keyExtractor'>;
  /**
   * PickerBackdropのProps
   */
  pickerBackdropProps?: Omit<PickerBackdropProps, 'isVisible' | 'onPress'>;
  /**
   * PickerContainerのProps
   */
  pickerContainerProps?: Omit<PickerContainerProps, 'isVisible'>;
  /**
   * PickerAccessoryのProps
   */
  pickerAccessoryProps?: Omit<DefaultPickerAccessoryProps, 'onDelete' | 'onCancel' | 'onDone'>;
  /**
   * アイテムのFontColor
   */
  itemColor?: string;
  /**
   * アイテムのFontFamily
   */
  itemFontFamily?: string;
};

const DEFAULT_DURATION = 300;

export const YearMonthPicker = (props: YearMonthPickerProps) => {
  const {
    isVisible,
    selectedYear,
    selectedMonth,
    yearItems,
    monthItems,
    onValueChangeYear,
    onValueChangeMonth,
    selectedLabel,
    open,
    handleBackdropPress,
    handleDelete,
    handleCancel,
    handleDone,
  } = useYearMonthPicker(props);
  const {
    yearSuffixLabel,
    monthSuffixLabel,
    placeholder,
    textInputProps,
    pickerItemsContainerProps: {style: pickerItemsContainerStyle, ...pickerItemsContainerProps} = {},
    pickerProps,
    pickerBackdropProps: {
      fadeInDuration = DEFAULT_DURATION,
      fadeOutDuration = DEFAULT_DURATION,
      ...pickerBackdropProps
    } = {},
    pickerContainerProps: {
      style: pickerContainerStyle,
      slideInDuration = DEFAULT_DURATION,
      slideOutDuration = DEFAULT_DURATION,
      ...pickerContainerProps
    } = {},
    pickerAccessoryProps,
  } = props;

  return (
    <>
      <PickerBackdrop
        isVisible={isVisible}
        onPress={handleBackdropPress}
        fadeInDuration={fadeInDuration}
        fadeOutDuration={fadeOutDuration}
        {...pickerBackdropProps}>
        <PickerContainer
          isVisible={isVisible}
          style={[styles.pickerContainer, pickerContainerStyle]}
          slideInDuration={slideInDuration}
          slideOutDuration={slideOutDuration}
          {...pickerContainerProps}>
          <DefaultPickerAccessory
            onDelete={handleDelete}
            onCancel={handleCancel}
            onDone={handleDone}
            {...pickerAccessoryProps}
          />
          <View
            style={StyleSheet.flatten([styles.pickerItemsContainer, pickerItemsContainerStyle])}
            {...pickerItemsContainerProps}>
            <View style={styles.flex}>
              <SelectPickerItems
                selectedValue={selectedYear}
                items={yearItems}
                onValueChange={onValueChangeYear}
                testID="yearPicker"
                {...pickerProps}
              />
            </View>
            <Text>{yearSuffixLabel}</Text>
            <View style={styles.flex}>
              <SelectPickerItems
                selectedValue={selectedMonth}
                items={monthItems}
                onValueChange={onValueChangeMonth}
                testID="monthPicker"
                {...pickerProps}
              />
            </View>
            <Text>{monthSuffixLabel}</Text>
          </View>
        </PickerContainer>
      </PickerBackdrop>
      <Pressable onPress={open} testID="pressableContainer">
        <View pointerEvents="box-only">
          <TextInput placeholder={placeholder} value={selectedLabel} editable={false} {...textInputProps} />
        </View>
      </Pressable>
    </>
  );
};

const styles = StyleSheet.create({
  flex: {
    flex: 1,
  },
  pickerContainer: {
    backgroundColor: 'white',
  },
  pickerItemsContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 10,
  },
});
