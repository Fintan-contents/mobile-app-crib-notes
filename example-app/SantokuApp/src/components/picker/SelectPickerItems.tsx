import {PickerProps} from '@react-native-picker/picker';
import {log} from 'framework/logging';
import React from 'react';
import {Platform} from 'react-native';

import {Item} from './SelectPicker';

export type SelectPickerItemsProps = PickerProps<React.Key> & {
  items: Item<React.Key>[];
  keyExtractor?: ((item: React.Key, index: number) => React.Key) | undefined;

  // android
  initialValue?: React.Key;
  itemHeight?: number;
  preferredNumVisibleRows?: number;
};

export const SelectPickerItems = (props: SelectPickerItemsProps) => {
  React.useEffect(() => {
    log.warn(`SelectPickerItems is not supported on: ${Platform.OS}`);
  }, []);
  return null;
};
