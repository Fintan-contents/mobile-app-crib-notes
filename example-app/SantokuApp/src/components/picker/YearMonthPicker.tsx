import {PickerProps} from '@react-native-picker/picker';
import {log} from 'framework/logging';
import React from 'react';
import {Platform, StyleProp, TextInputProps as RNETextInputProps, TextStyle, ViewProps} from 'react-native';

import {DefaultPickerAccessoryProps} from './DefaultPickerAccessory';
import {PickerBackdropProps} from './PickerBackdrop';
import {PickerContainerProps} from './PickerContainer';
import {Item} from './SelectPicker';
import {RequiredYearMonth, YearMonth} from './YearMonth';

type TextInputProps = Omit<RNETextInputProps, 'value' | 'editable'>;
export type YearMonthPickerProps = {
  /**
   * 選択された年月
   */
  selectedValue: YearMonth | undefined;
  /**
   * 選択可能な最大日付
   */
  maximumYearMonth: RequiredYearMonth;
  /**
   * 選択可能な最小日付
   */
  minimumYearMonth: RequiredYearMonth;
  /**
   * 年のサフィックス
   */
  yearSuffixLabel?: string;
  /**
   * 月のサフィックス
   */
  monthSuffixLabel?: string;
  /**
   * 未選択を表すアイテムのラベル
   * @platform android
   */
  unselectItem?: Item<number | undefined>;
  /**
   * アイテムが選択された場合に呼び出される関数
   */
  onSelectedItemChange?: (yearMonth?: YearMonth) => void;
  /**
   * PickerBackdropをタップして閉じた場合に呼び出される関数
   *
   * @platform ios
   */
  onDismiss?: (yearMonth?: YearMonth) => void;
  /**
   * DeleteLabelがタップされた場合に呼び出される関数
   * タップ後、YearMonthPickerは自動で閉じます。
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * @platform ios
   */
  onDelete?: (yearMonth?: YearMonth) => void;
  /**
   * CancelLabelがタップされた場合に呼び出される関数
   * タップ後、YearMonthPickerは自動で閉じます。
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * @platform ios
   */
  onCancel?: (yearMonth?: YearMonth) => void;
  /**
   * [iOS]
   * DoneLabelがタップされた場合に呼び出される関数
   * タップ後、YearMonthPickerは自動で閉じます。
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * [Android]
   * アイテムが選択された場合に呼び出される関数
   */
  onDone?: (yearMonth?: YearMonth) => void;
  /**
   * プレースホルダ
   */
  placeholder?: string;
  /**
   * 選択されたアイテムを表示するテキストコンポーネントのProps
   * なお、textInputComponentを指定した場合は使用されません。
   */
  textInputProps?: TextInputProps;
  /**
   * Pickerコンポーネントを囲むContainerのProps
   */
  pickerItemsContainerProps?: ViewProps;
  /**
   * PickerのProps
   * なお、pickerItemsComponentを指定した場合は使用されません。
   */
  pickerProps?: Omit<PickerProps<Date>, 'selectedValue' | 'onValueChange'>;
  /**
   * PickerBackdropのProps
   *
   * @platform ios
   */
  pickerBackdropProps?: Omit<PickerBackdropProps, 'isVisible' | 'onPress'>;
  /**
   * PickerContainerのProps
   *
   * @platform ios
   */
  pickerContainerProps?: Omit<PickerContainerProps, 'isVisible'>;
  /**
   * PickerAccessoryのProps
   * なお、pickerAccessoryComponentを指定した場合は使用されません。
   *
   * @platform ios
   */
  pickerAccessoryProps?: Omit<DefaultPickerAccessoryProps, 'onDelete' | 'onCancel' | 'onDone'>;
  /**
   * アイテムのFontColor
   */
  itemColor?: string;
  /**
   * アイテムのFontFamily
   */
  itemFontFamily?: string;
  /**
   * アイテムに指定するスタイル
   * 以下のプロパティのみ使用できます。
   * - 'color'
   * - 'backgroundColor'
   * - 'fontSize'
   * - 'fontFamily'
   *
   * @platform android
   */
  itemStyle?: StyleProp<Pick<TextStyle, 'color' | 'backgroundColor' | 'fontSize' | 'fontFamily'>>;
};

export const YearMonthPicker = (props: YearMonthPickerProps) => {
  React.useEffect(() => {
    log.warn(`YearMonthPicker is not supported on: ${Platform.OS}`);
  }, []);
  return null;
};
