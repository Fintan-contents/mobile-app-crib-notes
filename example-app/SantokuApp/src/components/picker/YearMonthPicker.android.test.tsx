import {render} from '@testing-library/react-native';
import React from 'react';
import {TextInputProps, ViewProps} from 'react-native';

import {ApplicationError} from '../../framework/error/ApplicationError';
import {SelectPickerItemsProps} from './SelectPickerItems';
import {YearMonthPicker} from './YearMonthPicker.android';
import {YearMonthUtil} from './YearMonthUtil';

describe('YearMonthPicker only with required props', () => {
  beforeAll(() => {
    const mockDate = new Date(2022, 3);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
  });
  afterAll(() => {
    jest.useRealTimers();
  });
  it('renders successfully', () => {
    const now = YearMonthUtil.now();
    const sut = render(
      <YearMonthPicker
        maximumYearMonth={now}
        minimumYearMonth={YearMonthUtil.addMonth(now, -60)}
        selectedValue={now}
        pickerItemsContainerProps={{testID: 'pickerItemsContainer'}}
      />,
    );

    expect(sut).toMatchSnapshot('YearMonthPicker only with required props.');

    const pickerItemsContainer = sut.queryByTestId('pickerItemsContainer');
    const yearTextInput = sut.queryByTestId('yearTextInput');
    const monthTextInput = sut.queryByTestId('monthTextInput');
    expect(pickerItemsContainer).not.toBeNull();
    expect(yearTextInput).not.toBeNull();
    expect(monthTextInput).not.toBeNull();

    const yearTextInputProps = yearTextInput?.props as TextInputProps;
    expect(yearTextInputProps.value).toBe('2022');
    const monthTextInputProps = monthTextInput?.props as TextInputProps;
    expect(monthTextInputProps.value).toBe('4');
  });
});

describe('YearMonthPicker with various maximum and minimum yearMonth', () => {
  it('same yearMonth', () => {
    const yearMonth = {year: 2022, month: 4};
    const sut = render(
      <YearMonthPicker maximumYearMonth={yearMonth} minimumYearMonth={yearMonth} selectedValue={undefined} />,
    );

    const yearPicker = sut.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([{value: 2022, label: '2022'}]);
    const monthPicker = sut.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([{value: 4, label: '4'}]);

    sut.update(
      <YearMonthPicker
        maximumYearMonth={yearMonth}
        minimumYearMonth={yearMonth}
        selectedValue={{year: yearMonth.year}}
      />,
    );

    const yearPicker2 = sut.getByTestId('yearPicker');
    const yearPickerProps2 = yearPicker2.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps2.items).toEqual([{value: 2022, label: '2022'}]);
    const monthPicker2 = sut.getByTestId('monthPicker');
    const monthPickerProps2 = monthPicker2.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps2.items).toEqual([{value: 4, label: '4'}]);
  });

  it('same year and different month', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2022, month: 1};
    const sut = render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={undefined}
      />,
    );

    const yearPicker = sut.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([{value: 2022, label: '2022'}]);
    const monthPicker = sut.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
    ]);

    sut.update(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={{year: maximumYearMonth.year}}
      />,
    );

    const yearPicker2 = sut.getByTestId('yearPicker');
    const yearPickerProps2 = yearPicker2.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps2.items).toEqual([{value: 2022, label: '2022'}]);
    const monthPicker2 = sut.getByTestId('monthPicker');
    const monthPickerProps2 = monthPicker2.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps2.items).toEqual([
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
    ]);
  });

  it('different year', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2021, month: 11};
    const sut = render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={undefined}
      />,
    );

    const yearPicker = sut.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([
      {value: 2021, label: '2021'},
      {value: 2022, label: '2022'},
    ]);
    const monthPicker = sut.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
      {value: 5, label: '5'},
      {value: 6, label: '6'},
      {value: 7, label: '7'},
      {value: 8, label: '8'},
      {value: 9, label: '9'},
      {value: 10, label: '10'},
      {value: 11, label: '11'},
      {value: 12, label: '12'},
    ]);

    // 最大年を選択
    sut.update(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={{year: maximumYearMonth.year}}
      />,
    );

    const yearPicker2 = sut.getByTestId('yearPicker');
    const yearPickerProps2 = yearPicker2.props as SelectPickerItemsProps<string>;

    expect(yearPickerProps2.items).toEqual([
      {value: 2021, label: '2021'},
      {value: 2022, label: '2022'},
    ]);

    const monthPicker2 = sut.getByTestId('monthPicker');
    const monthPickerProps2 = monthPicker2.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps2.items).toEqual([
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
    ]);

    // 最小年を選択
    sut.update(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={{year: minimumYearMonth.year}}
      />,
    );

    const yearPicker3 = sut.getByTestId('yearPicker');
    const yearPickerProps3 = yearPicker3.props as SelectPickerItemsProps<string>;

    expect(yearPickerProps3.items).toEqual([
      {value: 2021, label: '2021'},
      {value: 2022, label: '2022'},
    ]);

    const monthPicker3 = sut.getByTestId('monthPicker');
    const monthPickerProps3 = monthPicker3.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps3.items).toEqual([
      {value: 11, label: '11'},
      {value: 12, label: '12'},
    ]);
  });

  it('with unselectItem', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2022, month: 2};
    const sut = render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={maximumYearMonth}
        unselectItem={{value: undefined, label: 'please select...'}}
      />,
    );

    const yearPicker = sut.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([
      {value: undefined, label: 'please select...'},
      {value: 2022, label: '2022'},
    ]);
    const monthPicker = sut.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([
      {value: undefined, label: 'please select...'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
    ]);
  });

  it('maximum < minimum', () => {
    // コンポーネントマウント時のエラーログが出力されないようにモック化しておく
    jest.spyOn(console, 'error').mockImplementation();
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2022, month: 5};
    expect(() =>
      render(
        <YearMonthPicker
          maximumYearMonth={maximumYearMonth}
          minimumYearMonth={minimumYearMonth}
          selectedValue={maximumYearMonth}
        />,
      ),
    ).toThrow(new ApplicationError('Maximum yearMonth must be greater than or equal to minimum yearMonth.'));
  });
});

describe('YearMonthPicker with all props', () => {
  it('should be applied all properly', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const onSelectedItemChange = jest.fn();
    const onDone = jest.fn();
    /**
     * onDoneは自動テストではイベント発火できませんでした
     */
    const sut = render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={YearMonthUtil.addMonth(maximumYearMonth, -60)}
        unselectItem={{value: undefined, label: 'please select...'}}
        yearSuffixLabel="年"
        monthSuffixLabel="月"
        selectedValue={maximumYearMonth}
        onSelectedItemChange={onSelectedItemChange}
        placeholder="please select..."
        textInputProps={{style: {color: 'red'}}}
        pickerItemsContainerProps={{pointerEvents: 'none', testID: 'pickerItemsContainer'}}
        pickerProps={{numberOfLines: 5}}
        onDone={onDone}
        itemColor="yellow"
        itemFontFamily="Roboto"
        itemStyle={{backgroundColor: 'red'}}
      />,
    );

    expect(sut).toMatchSnapshot('YearMonthPicker all properly.');

    // assert pickerItemsContainer
    const pickerItemsContainer = sut.getByTestId('pickerItemsContainer');
    const pickerItemsContainerProps = pickerItemsContainer.props as ViewProps;
    expect(pickerItemsContainerProps.pointerEvents).toBe('none');

    // assert PickerItems
    const yearPicker = sut.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.numberOfLines).toBe(5);
    // itemsは、Itemの各Propを検証できなかったため、件数のみ検証する
    expect(yearPickerProps.items.length).toBe(7);
    const monthPicker = sut.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.numberOfLines).toBe(5);
    // itemsは、Itemの各Propを検証できなかったため、件数のみ検証する
    expect(monthPickerProps.items.length).toBe(5);
    const yearSuffix = sut.queryByText('年');
    const monthSuffix = sut.queryByText('月');
    expect(yearSuffix).not.toBeNull();
    expect(monthSuffix).not.toBeNull();

    // assert textInput
    const yearTextInput = sut.getByTestId('yearTextInput');
    const monthTextInput = sut.getByTestId('monthTextInput');
    const yearTextInputProps = yearTextInput.props as TextInputProps;
    expect(yearTextInputProps.style).toEqual({color: 'red'});
    expect(yearTextInputProps.value).toBe('2022');
    const monthTextInputProps = monthTextInput.props as TextInputProps;
    expect(monthTextInputProps.style).toEqual({color: 'red'});
    expect(monthTextInputProps.value).toBe('4');
  });
});
