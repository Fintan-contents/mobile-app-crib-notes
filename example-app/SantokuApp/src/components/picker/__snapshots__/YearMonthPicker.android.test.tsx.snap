// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearMonthPicker only with required props renders successfully: YearMonthPicker only with required props. 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingHorizontal": 10,
    }
  }
  testID="pickerItemsContainer"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={false}
      rejectResponderTermination={true}
      testID="yearTextInput"
      underlineColorAndroid="transparent"
      value="2022"
    />
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "transparent",
          "height": "100%",
          "opacity": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <RNCPicker
        items={
          Array [
            Object {
              "label": "2017",
              "testID": undefined,
              "textColor": undefined,
              "value": 2017,
            },
            Object {
              "label": "2018",
              "testID": undefined,
              "textColor": undefined,
              "value": 2018,
            },
            Object {
              "label": "2019",
              "testID": undefined,
              "textColor": undefined,
              "value": 2019,
            },
            Object {
              "label": "2020",
              "testID": undefined,
              "textColor": undefined,
              "value": 2020,
            },
            Object {
              "label": "2021",
              "testID": undefined,
              "textColor": undefined,
              "value": 2021,
            },
            Object {
              "label": "2022",
              "testID": undefined,
              "textColor": undefined,
              "value": 2022,
            },
          ]
        }
        numberOfLines={1}
        onChange={[Function]}
        selectedIndex={5}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
        testID="yearPicker"
      />
    </View>
  </View>
  <Text />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={false}
      rejectResponderTermination={true}
      testID="monthTextInput"
      underlineColorAndroid="transparent"
      value="4"
    />
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "transparent",
          "height": "100%",
          "opacity": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <RNCPicker
        items={
          Array [
            Object {
              "label": "1",
              "testID": undefined,
              "textColor": undefined,
              "value": 1,
            },
            Object {
              "label": "2",
              "testID": undefined,
              "textColor": undefined,
              "value": 2,
            },
            Object {
              "label": "3",
              "testID": undefined,
              "textColor": undefined,
              "value": 3,
            },
            Object {
              "label": "4",
              "testID": undefined,
              "textColor": undefined,
              "value": 4,
            },
          ]
        }
        numberOfLines={1}
        onChange={[Function]}
        selectedIndex={3}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
        testID="monthPicker"
      />
    </View>
  </View>
  <Text />
</View>
`;

exports[`YearMonthPicker with all props should be applied all properly: YearMonthPicker all properly. 1`] = `
<View
  pointerEvents="none"
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingHorizontal": 10,
    }
  }
  testID="pickerItemsContainer"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={false}
      placeholder="please select..."
      rejectResponderTermination={true}
      style={
        Object {
          "color": "red",
        }
      }
      testID="yearTextInput"
      underlineColorAndroid="transparent"
      value="2022"
    />
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "transparent",
          "height": "100%",
          "opacity": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <RNCPicker
        items={
          Array [
            Object {
              "label": "please select...",
              "testID": undefined,
              "textColor": undefined,
              "value": undefined,
            },
            Object {
              "label": "2017",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2017,
            },
            Object {
              "label": "2018",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2018,
            },
            Object {
              "label": "2019",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2019,
            },
            Object {
              "label": "2020",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2020,
            },
            Object {
              "label": "2021",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2021,
            },
            Object {
              "label": "2022",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2022,
            },
          ]
        }
        numberOfLines={5}
        onChange={[Function]}
        selectedIndex={6}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
        testID="yearPicker"
      />
    </View>
  </View>
  <Text>
    年
  </Text>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={false}
      placeholder="please select..."
      rejectResponderTermination={true}
      style={
        Object {
          "color": "red",
        }
      }
      testID="monthTextInput"
      underlineColorAndroid="transparent"
      value="4"
    />
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "transparent",
          "height": "100%",
          "opacity": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <RNCPicker
        items={
          Array [
            Object {
              "label": "please select...",
              "testID": undefined,
              "textColor": undefined,
              "value": undefined,
            },
            Object {
              "label": "1",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 1,
            },
            Object {
              "label": "2",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 2,
            },
            Object {
              "label": "3",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 3,
            },
            Object {
              "label": "4",
              "testID": undefined,
              "textColor": 4294967040,
              "value": 4,
            },
          ]
        }
        numberOfLines={5}
        onChange={[Function]}
        selectedIndex={4}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
        testID="monthPicker"
      />
    </View>
  </View>
  <Text>
    月
  </Text>
</View>
`;
